viewing_party/party.py

```diff
@@ -1,23 +1,177 @@
+    else:
+        movie_dict["title"] = title
+        movie_dict["genre"] = genre
+        movie_dict["rating"] = rating
+        return movie_dict
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    avg_rating = 0
+    for i in range(len(user_data["watched"])):
+        avg_rating += user_data["watched"][i]["rating"]
+    if len(user_data["watched"]) == 0:
+        return avg_rating
+    avg_rating /= len(user_data["watched"])
+    return avg_rating
+
+#second swing at get_watched_avg_rating. Which one is better?
+# def get_watched_avg_rating(user_data):
+#     user_watched = user_data["watched"]
+#     avg_rating = 0
+    
+#     for movie in user_watched:
+#         avg_rating += movie["rating"]
+#         print(avg_rating)
+#     if len(user_watched) == 0:
+#         return 0.0
+#     else:
+#         avg_rating /= len(user_watched)
+#         return avg_rating
+
+def get_most_watched_genre(user_data):
+    genre_dict = {}
+    i = 0
+    if len(user_data["watched"]) == 0:
+        return None
+    for key, value in user_data["watched"][i].items():
+        if key == "genre":
+            if value in genre_dict:
+                genre_dict[value] += 1
+            else:
+                genre_dict[value] = 1
+        i += 1
+    max_value = max(genre_dict.values())
+    for key, value in genre_dict.items():
+        if value == max_value:
+            return key
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+#returns unique movies that user has watched, but friends have not
+def get_unique_watched(user_data):
+    unique_watched_list = []
+
+    friend_watched_set = set(get_friends_watched(user_data))
+    user_watched_set = set(get_users_watched(user_data))
+
+    unique_set = user_watched_set - friend_watched_set
+
+    for movie in unique_set:
+        for i in range(len(user_data["watched"])):
+            if user_data["watched"][i]['title'] == movie:
+                unique_watched_list.append(user_data["watched"][i])   
+
+    return unique_watched_list
+    
+    
+    
+def get_friends_watched(user_data):   
+    friends_list_of_dict = user_data["friends"]
+    friends_watched_list = []
+
+    for i in range(len(friends_list_of_dict)):
+        for watched in friends_list_of_dict[i]:
+            for movie in friends_list_of_dict[i][watched]:

```

Comment:
There is a small potential bug here.  This second for loop is looping over all of the keys in a friends dictionary, and it works here because in our examples friends only have one key, `watched`.  Friends could theoretically have a "watchlist" or "subscriptions" or other data as part of the dictionary, and in those cases, this loop would pull in extra unwanted data.  However, even if we define a friend dictionary to only have the one key, this middle loop is not necessary.  The key-value pair for the 'watched' key can be accessed directly using square brackets:
```suggestion
    for i in range(len(friends_list_of_dict)):
        watched = friends_list_of_dict[i]["watched"]
        for movie in watched:

```